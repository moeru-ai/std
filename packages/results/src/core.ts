export interface Err<E> {
  __type__: 'err'
  error: E
}

export interface None {
  __type__: 'none'
}

export interface Ok<T> {
  __type__: 'ok'
  value: T
}

export type Option<T> = None | Some<T>

export type Result<T, E> = Err<E> | Ok<T>

export interface Some<T> {
  __type__: 'some'
  value: T
}

export const err = <T, E>(error: E): Result<T, E> => ({
  __type__: 'err',
  error,
})

export const none: Option<never> = {
  __type__: 'none',
}

export const ok = <T, E>(value: T): Result<T, E> => ({
  __type__: 'ok',
  value,
})

export const some = <T>(value: T): Option<T> => ({
  __type__: 'some',
  value,
})
