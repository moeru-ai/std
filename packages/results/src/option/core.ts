export interface None {
  __type__: 'none'
}

export type Option<T> = None | Some<T>

export interface Some<T> {
  __type__: 'some'
  value: T
}

export const none: Option<never> = {
  __type__: 'none',
}

export const some = <T>(value: T): Option<T> => ({
  __type__: 'some',
  value,
})
