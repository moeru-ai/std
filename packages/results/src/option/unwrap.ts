import type { Option } from '../core'

import { isSome } from './is'

export const unwrap = <T>(o: Option<T>): T => {
  if (isSome(o))
    return o.value
  else
    throw new Error('called `option.unwrap()` on a `None` value')
}

export const unwrapOr = <T>(o: Option<T>, fallback: T): T =>
  isSome(o)
    ? o.value
    : fallback

export const unwrapOrElse = <T>(o: Option<T>, onNone: () => T): T =>
  isSome(o)
    ? o.value
    : onNone()

// eslint-disable-next-line sonarjs/no-identical-functions
export const unwrapOrElseAsync = async <T>(o: Option<T>, onNone: () => Promise<T>): Promise<T> =>
  isSome(o)
    ? o.value
    : onNone()
