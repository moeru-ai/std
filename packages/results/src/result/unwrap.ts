import type { Result } from './core'

import { isErr, isOk } from './is'

export const unwrap = <T, E>(r: Result<T, E>): T => {
  if (isOk(r))
    return r.value
  else if (r.error instanceof Error)
    throw r.error
  else
    throw new Error('called `result.unwrap()` on an `Err` value', { cause: r.error })
}

export const unwrapErr = <T, E>(r: Result<T, E>): E => {
  if (isErr(r))
    return r.error
  else
    throw new Error('called `result.unwrapErr()` on an `Ok` value', { cause: r.value })
}

export const unwrapOr = <T, E>(r: Result<T, E>, fallback: T): T =>
  isOk(r)
    ? r.value
    : fallback

export const unwrapOrElse = <T, E>(r: Result<T, E>, onErrValue: (e: E) => T): T =>
  isOk(r)
    ? r.value
    : onErrValue(r.error)

// eslint-disable-next-line sonarjs/no-identical-functions
export const unwrapOrElseAsync = async <T, E>(r: Result<T, E>, onErrValue: (e: E) => Promise<T>): Promise<T> =>
  isOk(r)
    ? r.value
    : onErrValue(r.error)
