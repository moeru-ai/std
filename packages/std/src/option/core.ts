export interface None {
  __mo: 'none'
}

export type Option<T> = None | Some<T>

export interface Some<T> {
  __mo: 'some'
  value: T
}

export const none: Option<never> = {
  __mo: 'none',
}

export const some = <T>(value: T): Option<T> => ({
  __mo: 'some',
  value,
})
