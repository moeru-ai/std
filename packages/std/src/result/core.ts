export interface Err<E> {
  __mr: 'err'
  error: E
}

export interface Ok<T> {
  __mr: 'ok'
  value: T
}

export type Result<T, E> = Err<E> | Ok<T>

export const err = <T, E>(error: E): Result<T, E> => ({
  __mr: 'err',
  error,
})

export const ok = <T, E>(value: T): Result<T, E> => ({
  __mr: 'ok',
  value,
})
