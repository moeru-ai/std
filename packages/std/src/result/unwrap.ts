import type { Result } from './core'

import { isOk, isErr } from './is'

export const unwrap = <T, E>(r: Result<T, E>): T => {
  if (isOk(r))
    return r.value
  else
    throw r.error
}

export const unwrapErr = <T, E>(r: Result<T, E>): E => {
  if (isErr(r))
    return r.error
  else
    throw new Error('called `result.unwrapErr()` on a `Ok` value', { cause: r.value })
}

export const unwrapOr = <T, E>(r: Result<T, E>, fallback: T): T =>
  isOk(r)
    ? r.value
    : fallback

export const unwrapOrElse = <T, E>(r: Result<T, E>, onErrValue: (e: E) => T): T =>
  isOk(r)
    ? r.value
    : onErrValue(r.error)

export const unwrapOrElseAsync = async <T, E>(r: Result<T, E>, onErrValue: (e: E) => Promise<T>): Promise<T> =>
  isOk(r)
    ? r.value
    : await onErrValue(r.error)
