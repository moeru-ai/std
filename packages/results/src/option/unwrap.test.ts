import { describe, expect, it } from 'vitest'

import { none, some } from '../core'
import { unwrap, unwrapOr, unwrapOrElse } from './unwrap'

describe('@moeru/results', () => {
  // https://doc.rust-lang.org/std/option/enum.Option.html#examples-9
  it('option.unwrap', () => {
    expect(unwrap(some('air'))).toBe('air')
    expect(() => unwrap(none)).toThrowErrorMatchingSnapshot()
  })

  // https://doc.rust-lang.org/std/option/enum.Option.html#examples-10
  it('option.unwrapOr', () => {
    expect(unwrapOr(some('car'), 'bike')).toBe('car')
    expect(unwrapOr(none, 'bike')).toBe('bike')
  })

  // https://doc.rust-lang.org/std/option/enum.Option.html#examples-11
  it('option.unwrapOrElse', () => {
    expect(unwrapOrElse(some(4), () => 20)).toBe(4)
    expect(unwrapOrElse(none, () => 20)).toBe(20)
  })
})
