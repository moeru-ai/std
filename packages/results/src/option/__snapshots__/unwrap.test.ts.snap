// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@moeru/results > option.unwrap 1`] = `[Error: called \`option.unwrap()\` on a \`None\` value]`;
